internal/controller/postgrescluster/patroni.go

```diff
@@ -499,6 +512,44 @@ func (r *Reconciler) reconcilePatroniSwitchover(ctx context.Context,
 			stdout, stderr, command...)
 	}
 
+	// To ensure idempotency, the operator verifies that the timeline reported by Patroni
+	// matches the timeline that was present when the switchover was first requested.
+	// TODO(benjb): consider pulling the timeline from the pod annotation; manual experiments
+	// have shown that the annotation on the Leader pod is up to date during a switchover, but
+	// missing from the Replica pods.
+	timeline, err := patroni.Executor(exec).GetTimeline(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	if timeline == 0 {
+		return errors.New("error getting and parsing current timeline")
+	}
+
+	statusTimeline := cluster.Status.Patroni.ExpectedTimeline
+
+	// If the `ExpectedTimeline` field is empty, this is the first reconcile after
+	// a switchover has been requested and we need to fill in the field with the current TL
+	// as reported by Patroni.
+	// We return from here without calling for an explicit requeue, but since we're updating
+	// the object, we will reconcile this again for the actual switchover/failover action.
+	if statusTimeline == nil || (statusTimeline != nil && *statusTimeline == 0) {
+		log.V(1).Info("Setting ExpectedTimeline to current timeline")
+		cluster.Status.Patroni.ExpectedTimeline = &timeline
+		return err
+	}
+
+	// If the `ExpectedTimeline` field does not match the current timeline as reported by Patroni,
+	// then we assume a switchover has been completed, and we have reached this point because the
+	// cache does not yet have the updated `cluster.Status.Patroni.Switchover` field.
```

Comment:
This is the part that I am most leery about, and it's really around that "assume", but like you point out (and considering that timelines are increasing, like we talked about yesterday):

(a) a user annotates a cluster; 
(b) we get the timeline from patroni and add that to the cluster; 
(c) next time we check patroni, if it doesn't match what have on the cluster, e.g., what we we got from patroni at a previous time, then a timeline increase has happened between (b) and (c), which means a switchover _of some sort_ has occurred, yeah?

So we know a switchover occurred between the user asking for one and now; whether that timeline change was us, however, is not entirely clear in this situation. That's the part that I'm most worried about, but that's also the part that I think we have to assume for now because if we don't, then we're back at "single triggers causing multiple failovers."

Reply:
> then a timeline increase has happened between (b) and (c), which means a switchover _of some sort_ has occurred, yeah?

That's my thinking, right.

> So we know a switchover occurred between the user asking for one and now; whether that timeline change was us, however, is not entirely clear in this situation. That's the part that I'm most worried about, but that's also the part that I think we have to assume for now because if we don't, then we're back at "single triggers causing multiple failover."

The fields on the spec describe a desire to change the leader of the cluster. I don't think it matters what _caused_ the leader to change; it only matters that it changed. We're altering the criteria from "we called patronictl" to "we see a Postgres promotion."

That's where it gets fuzzy, I suppose. We still have "if called successfully, set done" and I'm not opposed to keeping that (especially in a bug fix). One of those criteria/transitions is the "main" one and the other is some approximation. I can't decide now which deserves what role. No changes are needed, IMO.

Reply:
I was thinking about adding something to the documentation about "always check your instance to make sure the post-fail/switchover state is what you want it to be" but that's already there with comments about checking the labels after a fail/switchover. 

(And I've tried to add code comments about the logic of the behavior, but is it enough?)